File: test-suite-kotlin/src/test/kotlin/io/micronaut/servlet/docs/servletannotation/FilterAttributeController.kt
package io.micronaut.servlet.docs.servletannotation

import io.micronaut.context.annotation.Requires
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import jakarta.servlet.http.HttpServletRequest

/**
 * Reports whether the filter registered by [MyFilterFactory] ran for the request, so the
 * documented mappings and order can be asserted.
 */
@Requires(property = "spec.name", value = "MyFilterFactoryTest")
@Controller
class FilterAttributeController {

    @Get(value = "/extra-filter/attribute", produces = [MediaType.TEXT_PLAIN])
    fun extraFilter(request: HttpServletRequest): String = java.lang.String.valueOf(request.getAttribute("runFirst"))

    @Get(value = "/mapped-filter/attribute", produces = [MediaType.TEXT_PLAIN])
    fun mappedFilter(request: HttpServletRequest): String = java.lang.String.valueOf(request.getAttribute("runFirst"))

    @Get(value = "/unfiltered/attribute", produces = [MediaType.TEXT_PLAIN])
    fun unfiltered(request: HttpServletRequest): String = java.lang.String.valueOf(request.getAttribute("runFirst"))
}

File: test-suite-python/src/test/python/micronaut/servlet/docs/servletannotation/FilterAttributeController.py
from jakarta.servlet.http import HttpServletRequest
from micronaut.context.annotation import Requires
from micronaut.http import MediaType
from micronaut.http.annotation import Controller, Get


def attribute(request: HttpServletRequest) -> str:
    value = request.getAttribute("runFirst")
    return "null" if value is None else str(value).lower()


# Reports whether the filter registered by MyFilterFactory ran for the request, so the
# documented mappings and order can be asserted.
@Requires(property="spec.name", value="MyFilterFactoryTest")
@Controller
class FilterAttributeController:

    @Get(value="/extra-filter/attribute", produces=MediaType.TEXT_PLAIN)
    def extra_filter(self, request: HttpServletRequest) -> str:
        return attribute(request)

    @Get(value="/mapped-filter/attribute", produces=MediaType.TEXT_PLAIN)
    def mapped_filter(self, request: HttpServletRequest) -> str:
        return attribute(request)

    @Get(value="/unfiltered/attribute", produces=MediaType.TEXT_PLAIN)
    def unfiltered(self, request: HttpServletRequest) -> str:
        return attribute(request)
